pkg/controller/ephemeralcluster/reconciler.go

```diff
@@ -392,43 +490,103 @@ func (r *reconciler) makeProwJob(ciOperatorConfig *api.ReleaseBuildConfiguration
 	return &pj, nil
 }
 
-func (r *reconciler) fetchKubeconfig(ctx context.Context, log *logrus.Entry, ecStatus *ephemeralclusterv1.EphemeralClusterStatus, pj *prowv1.ProwJob) error {
+func (r *reconciler) fetchSecrets(ctx context.Context, log *logrus.Entry, ec *ephemeralclusterv1.EphemeralCluster, ecStatus *ephemeralclusterv1.EphemeralClusterStatus, pj *prowv1.ProwJob) error {
 	buildClient, err := r.buildClientFor(pj)
 	if err != nil {
 		log.WithField("cluster", pj.Spec.Cluster).WithError(err).Error("Build client not found")
-		upsertCondition(ecStatus, ephemeralclusterv1.ClusterReady, ephemeralclusterv1.ConditionFalse, r.now(), ephemeralclusterv1.KubeconfigFetchFailureReason, err.Error())
+		upsertCondition(ecStatus, ephemeralclusterv1.ClusterReady, ephemeralclusterv1.ConditionFalse, r.now(), ephemeralclusterv1.SecretsFetchFailureReason, err.Error())
 		ecStatus.Phase = ephemeralclusterv1.EphemeralClusterFailed
 		return reconcile.TerminalError(err)
 	}
 
 	ns, err := r.findCIOperatorTestNS(ctx, buildClient, pj)
 	if err != nil {
 		log.Info("ci-operator NS not found")
-		upsertCondition(ecStatus, ephemeralclusterv1.ClusterReady, ephemeralclusterv1.ConditionFalse, r.now(), ephemeralclusterv1.KubeconfigFetchFailureReason, ephemeralclusterv1.CIOperatorNSNotFoundMsg)
+		upsertCondition(ecStatus, ephemeralclusterv1.ClusterReady, ephemeralclusterv1.ConditionFalse, r.now(), ephemeralclusterv1.SecretsFetchFailureReason, ephemeralclusterv1.CIOperatorNSNotFoundMsg)
 		return nil
 	}
 	log.WithField("namespace", ns).Info("ci-operator namespace found")
 
+	if ec.Spec.CIOperator.Test.ClusterClaim != nil {
+		r.fetchHiveSecrets(ctx, log, buildClient, ns, ecStatus)
+	} else {
+		r.fetchClusterKubeconfig(ctx, log, buildClient, ns, ecStatus)
+	}
+
+	return nil
+}
+
+func (r *reconciler) fetchHiveSecrets(ctx context.Context, log *logrus.Entry, buildClient ctrlruntimeclient.Client, ns string, ecStatus *ephemeralclusterv1.EphemeralClusterStatus) {
+	secretsReady := true
+	var err error
+	var kubeconfig, passwd []byte
+
+	for _, secret := range []struct {
+		name string
+		key  string
+		data *[]byte
+	}{
+		{name: HiveKubeconfigSecret, key: api.HiveAdminKubeconfigSecretKey, data: &kubeconfig},
+		{name: HiveAdminPasswdSecret, key: api.HiveAdminPasswordSecretKey, data: &passwd},
+	} {
+		s := corev1.Secret{}
+		if err = buildClient.Get(ctx, types.NamespacedName{Name: secret.name, Namespace: ns}, &s); err != nil {
+			secretsReady = false
+			if kerrors.IsNotFound(err) {
+				// Not available yet, do not report this as an error
+				err = nil
+			} else {
+				log.WithField("secret", secret.name).WithError(err).Error("Failed to read secret")
+			}
+			break
+		}
+
+		data, ok := s.Data[secret.key]
+		if !ok {
+			secretsReady = false
+			break
+		}
+
+		*secret.data = slices.Clone(data)
```

Comment:
This looks dangerous.

Reply:
But as the secret is just an element of the anonymous struct, you are trying to mutate it here and you are not using it anywhere else.

Reply:
`data` is a pointer to `kubeconfig` on the first iteration and to `passwd` on the second. Doing so on the first iteration:
```golang
*secret.data = slices.Clone(data)
```
is like:
```
kubeconfig = slices.Clone(data)
```
Test this out:
```golang
func TestPointerToSlice(t *testing.T) {
	var kubeconfig []byte
	var adminPasswd []byte

	for i, testCase := range []struct {
		data *[]byte
	}{
		{data: &kubeconfig},
		{data: &adminPasswd},
	} {
		*testCase.data = []byte(fmt.Sprintf("data-%d", i))
	}

	if string(kubeconfig) != "data-0" {
		t.Fatalf("want data-1 but got %s", kubeconfig)
	}

	if string(adminPasswd) != "data-1" {
		t.Fatalf("want data-2 but got %s", adminPasswd)
	}
}
```